File: apps/web/src/features/game/lib/ai/ev.ts
/**
 * Expected-value rule over candidate subsets.
 *
 * EV(S) = p(|S|) * mean_{c in S} Q(c) + (1 - p(|S|)) * Q_pass
 *
 * Because the placed cell is uniform over S, the optimal subset of size k is
 * exactly the k cells with the highest Q, so move choice reduces to comparing
 * top-k prefixes for k = 1..5. Line-for-line mirror of ml mcts/ev.py.
 */

import { MAX_CANDIDATES, SUCCESS_PROBABILITY } from "@pkg/shared/constants";

export interface SubsetChoice {
  /** Indices into the q array, best-first. */
  order: number[];
  ev: number;
}

/** EV of the top-k prefix for k = 1..min(maxK, q.length); q must be sorted descending. */
export function evCurve(
  qSortedDesc: number[],
  qPass: number,
  maxK: number = MAX_CANDIDATES,
): number[] {
  const limit = Math.min(maxK, qSortedDesc.length, MAX_CANDIDATES);
  const curve: number[] = [];
  let sum = 0;
  for (let k = 1; k <= limit; k++) {
    const q = qSortedDesc[k - 1];
    const probability = SUCCESS_PROBABILITY[k];
    if (q === undefined || probability === undefined) break;
    sum += q;
    curve.push(probability * (sum / k) + (1 - probability) * qPass);
  }
  return curve;
}

/** Returns the optimal subset (indices into qValues, best-first) and its EV. */
export function bestSubset(
  qValues: number[],
  qPass: number,
  maxK: number = MAX_CANDIDATES,
): SubsetChoice {
  if (qValues.length === 0) {
    throw new Error("bestSubset needs at least one q value");
  }
  const order = qValues
    .map((q, index) => ({ q, index }))
    .sort((a, b) => b.q - a.q || a.index - b.index)
    .map((entry) => entry.index);
  const sorted = order.map((index) => qValues[index] as number);
  const curve = evCurve(sorted, qPass, maxK);
  let bestK = 1;
  let bestEv = -Infinity;
  for (let k = 1; k <= curve.length; k++) {
    const ev = curve[k - 1];
    if (ev !== undefined && ev > bestEv) {
      bestEv = ev;
      bestK = k;
    }
  }
  return { order: order.slice(0, bestK), ev: bestEv };
}

File: ml/src/ungomoku_ml/mcts/ev.py
"""Expected-value rule over candidate subsets.

For a subset S of k cells with per-cell values Q(c) and failure value Q_pass:

    EV(S) = p(k) * mean_{c in S} Q(c) + (1 - p(k)) * Q_pass

Because the placed cell is uniform over S, for a fixed k the optimal subset is
exactly the k cells with the highest Q. Choosing a move therefore reduces to
comparing the EVs of the top-k prefixes for k = 1..5.

Mirrored line-for-line by the TS engine (apps/web .../lib/ai/ev.ts).
"""

import numpy as np

from ungomoku_ml.rules import MAX_CANDIDATES, SUCCESS_PROBABILITY

_PROBS = np.array([SUCCESS_PROBABILITY[k] for k in range(1, MAX_CANDIDATES + 1)])


def ev_curve(q_sorted_desc: np.ndarray, q_pass: float, max_k: int = MAX_CANDIDATES) -> np.ndarray:
    """EV of the top-k prefix for k = 1..min(max_k, len(q)).

    ``q_sorted_desc`` must be sorted in descending order.
    """
    limit = min(max_k, len(q_sorted_desc), MAX_CANDIDATES)
    if limit == 0:
        return np.empty(0, dtype=np.float64)
    prefix = q_sorted_desc[:limit]
    means = np.cumsum(prefix) / np.arange(1, limit + 1)
    probs = _PROBS[:limit]
    return probs * means + (1.0 - probs) * q_pass


def best_subset(
    q_values: np.ndarray,
    q_pass: float,
    max_k: int = MAX_CANDIDATES,
) -> tuple[np.ndarray, float]:
    """Returns (indices into q_values ordered best-first, EV) of the optimal subset."""
    order = np.argsort(-q_values, kind="stable")
    evs = ev_curve(q_values[order], q_pass, max_k)
    if len(evs) == 0:
        raise ValueError("best_subset needs at least one q value")
    k = int(np.argmax(evs)) + 1
    return order[:k], float(evs[k - 1])
